### ISSUE TITLE
Usage together with json_serializable

### ISSUE THREAD

**Issue by mvarendorff** [heftapp/graphql_codegen#297]
I am currently in an updating session for one of our projects which also included updating grapqhl_codegen from json_serializable times (0.10.x) to the current 0.13.4. Unfortunately, this broke two instances where we are using a generated type in a new class that's tagged with `@JsonSerializable()`. Using in this case means either of:

1) Extending a generated type (this is also not possible anymore because the generated constructor is a factory constructor which may not be extended; I can open a new issue for this if needed).
```dart
@JsonSerializable()
class Example extends Query$something {
  Example({required super.id});

  factory Example.fromJson(Map<String, dynamic> json) =>_$ExampleFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$ExampleToJson(this);
}
```

<details>
<summary>This fails with:</summary>

```
Cannot populate the required constructor argument: id.
package:example/example.dart:9:3
  ╷
9 │   Example({required super.id});
  │   ^^^^^^^^^^^^
```

</details>

---

2) Using a generated type in a field
```dart
@JsonSerializable()
class Example {
  final Query$something? value;

  Example({required this.value});

  factory Example.fromJson(Map<String, dynamic> json) =>_$ExampleFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$ExampleToJson(this);
}
```

<details>
<summary>This fails with:</summary>

```
Could not generate `fromJson` code for `value`.
To support the type `InvalidType` you can:
* Use `JsonConverter`
  https://pub.dev/documentation/json_annotation/latest/json_annotation/JsonConverter-class.html
* Use `JsonKey` fields `fromJson` and `toJson`
  https://pub.dev/documentation/json_annotation/latest/json_annotation/JsonKey/fromJson.html
  https://pub.dev/documentation/json_annotation/latest/json_annotation/JsonKey/toJson.html
package:nebolus/common/entities/fuckery.dart:10:46
   ╷
10 │   final Query$something? value;
   │                                              ^^^^^
   ╵
```

</details>

*The error persists, even when adding*
```dart
valueToJson(Query$something value) => value.toJson();
// snip

  @JsonKey(
    fromJson: Query$something.fromJson,
    toJson: valueToJson,
  )
```

---

Now I am curious: Is there any way to avoid these problems? I don't know if this is connected to the order of the runners perhaps but I also tried adding `runs_before: ["json_serializable", "json_serializable:json_serializable"]` to the build options for `graphql_codegen` in my `build.yaml` but to no avaail.

One option would obviously be to keep the older version around but there are so many things attached to it that would be held back as well that I don't really consider this a *good* option.

**Comment by budde377** [comment]
Thanks for creating this issue @mvarendorff. Honestly, I have no idea why the second example doesn't work or if this is caused by the order of the builders. 

I am curious, what kind of benefits are you getting from extending the query class?

**Comment by mvarendorff** [comment]
That is a very valid question! Turns out... None, really. The snippet that did it was written many years ago without much experience and there wasn't anything there that couldn't be handled by an extension instead. I believe the original idea was to enrich the object from the server with additional data from the app and then write that to disk to be retrieved later. In our case that data was also included in the server object already, so there wasn't really a point in doing what we were doing but if that wasn't the case, it would probably make more sense to store them separately and join them in code rather than in storage.

**Comment by mvarendorff** [comment]
As for the other, I did some digging and testing and trying. It turns out, the other was indeed the culprit as json_serializable by default appears to run before graphql_codegen.

Adding
```yaml
    runs_before:
      - json_serializable:json_serializable
```
to graphql_codegen's build.yaml and removing the custom `@JsonKey` made things work. I opened a PR at #299 which in my tests solves both portions of the issue described originally (eventhough the former has been worked around otherwise).
